File: node-api/src/services/storage.js
/**
 * NovelGrab – Object Storage Service
 *
 * Abstraction layer over S3-compatible storage (AWS S3, Cloudflare R2, MinIO)
 * with local filesystem fallback for development.
 *
 * Deterministic paths:
 *   /novels/{slug}/chapters/{chapterNumber}.json
 *   /novels/{slug}/artifact/book.epub
 *   /covers/{slug}.jpg
 *
 * Pattern from file-uploads skill:
 *   - S3 client with presigned URLs
 *   - Local fallback for dev
 *   - Content-type detection
 */

import fs from "fs/promises";
import path from "path";
import { fileURLToPath } from "url";
import config from "../config/index.js";
import { logger } from "../utils/logger.js";

const __dirname = path.dirname(fileURLToPath(import.meta.url));

// ─── S3 client (lazy-loaded) ──────────────────────────────────────

let s3Client = null;

async function getS3Client() {
  if (s3Client) return s3Client;

  const { S3Client } = await import("@aws-sdk/client-s3");
  s3Client = new S3Client({
    region: config.storage.s3.region,
    endpoint: config.storage.s3.endpoint || undefined,
    credentials: {
      accessKeyId: config.storage.s3.accessKeyId,
      secretAccessKey: config.storage.s3.secretAccessKey,
    },
    forcePathStyle: true, // Required for MinIO/R2
  });

  return s3Client;
}

// ─── Local Storage Helpers ─────────────────────────────────────────

function localStoragePath(key) {
  return path.resolve(config.storage.localPath, key);
}

async function ensureLocalDir(filePath) {
  const dir = path.dirname(filePath);
  await fs.mkdir(dir, { recursive: true });
}

// ─── Public API ────────────────────────────────────────────────────

/**
 * Upload content to object storage.
 * @param {string} key – Storage key (e.g., "novels/slug/chapters/0001.json")
 * @param {Buffer|string} content – File content
 * @param {string} contentType – MIME type
 * @returns {string} Public URL or local path
 */
export async function uploadObject(
  key,
  content,
  contentType = "application/json",
) {
  if (config.storage.provider === "s3") {
    return uploadToS3(key, content, contentType);
  }
  return uploadToLocal(key, content);
}

/**
 * Download content from object storage.
 * @param {string} key – Storage key
 * @returns {Buffer|null} Content or null if not found
 */
export async function downloadObject(key) {
  if (config.storage.provider === "s3") {
    return downloadFromS3(key);
  }
  return downloadFromLocal(key);
}

/**
 * Check if an object exists.
 * @param {string} key – Storage key
 * @returns {boolean}
 */
export async function objectExists(key) {
  if (config.storage.provider === "s3") {
    return existsInS3(key);
  }
  return existsInLocal(key);
}

/**
 * Delete an object.
 * @param {string} key – Storage key
 */
export async function deleteObject(key) {
  if (config.storage.provider === "s3") {
    return deleteFromS3(key);
  }
  return deleteFromLocal(key);
}

/**
 * Get a public/download URL for an object.
 * For S3: generates a presigned URL (1 hour).
 * For local: returns the API serve path.
 */
export async function getObjectUrl(key) {
  if (config.storage.provider === "s3") {
    return getS3PresignedUrl(key);
  }
  // Local: serve via API endpoint
  return `/api/storage/${key}`;
}

// ─── Deterministic Path Builders ───────────────────────────────────

export function chapterStorageKey(slug, chapterNumber) {
  const num = String(chapterNumber).padStart(4, "0");
  return `novels/${slug}/chapters/${num}.json`;
}

export function artifactStorageKey(slug) {
  return `novels/${slug}/artifact/book.epub`;
}

export function coverStorageKey(slug) {
  return `covers/${slug}.jpg`;
}

// ─── S3 Implementation ────────────────────────────────────────────

async function uploadToS3(key, content, contentType) {
  const { PutObjectCommand } = await import("@aws-sdk/client-s3");
  const client = await getS3Client();

  const body = typeof content === "string" ? Buffer.from(content) : content;

  await client.send(
    new PutObjectCommand({
      Bucket: config.storage.s3.bucket,
      Key: key,
      Body: body,
      ContentType: contentType,
    }),
  );

  logger.debug({ key }, "Uploaded to S3");

  if (config.cdnBaseUrl) {
    return `${config.cdnBaseUrl}/${key}`;
  }
  return `${config.storage.s3.endpoint}/${config.storage.s3.bucket}/${key}`;
}

async function downloadFromS3(key) {
  const { GetObjectCommand } = await import("@aws-sdk/client-s3");
  const client = await getS3Client();

  try {
    const resp = await client.send(
      new GetObjectCommand({
        Bucket: config.storage.s3.bucket,
        Key: key,
      }),
    );
    const chunks = [];
    for await (const chunk of resp.Body) {
      chunks.push(chunk);
    }
    return Buffer.concat(chunks);
  } catch (err) {
    if (err.name === "NoSuchKey" || err.$metadata?.httpStatusCode === 404) {
      return null;
    }
    throw err;
  }
}

async function existsInS3(key) {
  const { HeadObjectCommand } = await import("@aws-sdk/client-s3");
  const client = await getS3Client();

  try {
    await client.send(
      new HeadObjectCommand({
        Bucket: config.storage.s3.bucket,
        Key: key,
      }),
    );
    return true;
  } catch {
    return false;
  }
}

async function deleteFromS3(key) {
  const { DeleteObjectCommand } = await import("@aws-sdk/client-s3");
  const client = await getS3Client();

  await client.send(
    new DeleteObjectCommand({
      Bucket: config.storage.s3.bucket,
      Key: key,
    }),
  );
}

async function getS3PresignedUrl(key) {
  const { GetObjectCommand } = await import("@aws-sdk/client-s3");
  const { getSignedUrl } = await import("@aws-sdk/s3-request-presigner");
  const client = await getS3Client();

  const url = await getSignedUrl(
    client,
    new GetObjectCommand({
      Bucket: config.storage.s3.bucket,
      Key: key,
    }),
    { expiresIn: 3600 }, // 1 hour
  );

  return url;
}

// ─── Local Implementation ──────────────────────────────────────────

async function uploadToLocal(key, content) {
  const filePath = localStoragePath(key);
  await ensureLocalDir(filePath);

  const data = typeof content === "string" ? content : content;
  await fs.writeFile(filePath, data);

  logger.debug({ key, filePath }, "Uploaded to local storage");
  return `/api/storage/${key}`;
}

async function downloadFromLocal(key) {
  const filePath = localStoragePath(key);
  try {
    return await fs.readFile(filePath);
  } catch (err) {
    if (err.code === "ENOENT") return null;
    throw err;
  }
}

async function existsInLocal(key) {
  const filePath = localStoragePath(key);
  try {
    await fs.access(filePath);
    return true;
  } catch {
    return false;
  }
}

async function deleteFromLocal(key) {
  const filePath = localStoragePath(key);
  try {
    await fs.unlink(filePath);
  } catch (err) {
    if (err.code !== "ENOENT") throw err;
  }
}

File: python-worker/src/storage.py
"""
NovelGrab – Storage Client (Python)

Mirrors the Node.js storage service. Uploads chapter content
to local filesystem or S3-compatible storage.

Deterministic paths:
  /novels/{slug}/chapters/{chapterNumber}.json
"""

import os
import json
import logging
from pathlib import Path
from typing import Optional

from config import config

log = logging.getLogger("novelgrab.storage")


def chapter_storage_key(slug: str, chapter_number: int) -> str:
    """Build deterministic storage key for a chapter."""
    num = str(chapter_number).zfill(4)
    return f"novels/{slug}/chapters/{num}.json"


def _local_path(key: str) -> Path:
    """Resolve a storage key to a local filesystem path."""
    return Path(config.LOCAL_STORAGE_PATH) / key


# ─── Local Storage ──────────────────────────────────────────────────


def upload_local(key: str, content: str) -> str:
    """Write content to local storage. Returns the storage path."""
    file_path = _local_path(key)
    file_path.parent.mkdir(parents=True, exist_ok=True)
    file_path.write_text(content, encoding="utf-8")
    log.debug("Uploaded to local: %s", key)
    return str(file_path)


def download_local(key: str) -> Optional[str]:
    """Read content from local storage. Returns None if not found."""
    file_path = _local_path(key)
    if not file_path.exists():
        return None
    return file_path.read_text(encoding="utf-8")


def exists_local(key: str) -> bool:
    """Check if a file exists in local storage."""
    return _local_path(key).exists()


# ─── S3-Compatible Storage ──────────────────────────────────────────

_s3_client = None


def _get_s3_client():
    """Lazy-initialize S3 client."""
    global _s3_client
    if _s3_client is not None:
        return _s3_client

    import boto3

    kwargs = {
        "region_name": config.S3_REGION,
        "aws_access_key_id": config.S3_ACCESS_KEY_ID,
        "aws_secret_access_key": config.S3_SECRET_ACCESS_KEY,
    }
    if config.S3_ENDPOINT:
        kwargs["endpoint_url"] = config.S3_ENDPOINT

    _s3_client = boto3.client("s3", **kwargs)
    return _s3_client


def upload_s3(key: str, content: str, content_type: str = "application/json") -> str:
    """Upload content to S3-compatible storage."""
    client = _get_s3_client()
    client.put_object(
        Bucket=config.S3_BUCKET,
        Key=key,
        Body=content.encode("utf-8"),
        ContentType=content_type,
    )
    log.debug("Uploaded to S3: %s", key)
    return f"s3://{config.S3_BUCKET}/{key}"


def download_s3(key: str) -> Optional[str]:
    """Download content from S3-compatible storage."""
    client = _get_s3_client()
    try:
        resp = client.get_object(Bucket=config.S3_BUCKET, Key=key)
        return resp["Body"].read().decode("utf-8")
    except client.exceptions.NoSuchKey:
        return None
    except Exception:
        return None


def exists_s3(key: str) -> bool:
    """Check if an object exists in S3."""
    client = _get_s3_client()
    try:
        client.head_object(Bucket=config.S3_BUCKET, Key=key)
        return True
    except Exception:
        return False


# ─── Unified Interface ─────────────────────────────────────────────


def upload(key: str, content: str, content_type: str = "application/json") -> str:
    """Upload content to configured storage provider."""
    if config.STORAGE_PROVIDER == "s3":
        return upload_s3(key, content, content_type)
    return upload_local(key, content)


def download(key: str) -> Optional[str]:
    """Download content from configured storage provider."""
    if config.STORAGE_PROVIDER == "s3":
        return download_s3(key)
    return download_local(key)


def exists(key: str) -> bool:
    """Check if content exists in configured storage provider."""
    if config.STORAGE_PROVIDER == "s3":
        return exists_s3(key)
    return exists_local(key)
